app/Services/StandService.php

```diff
@@ -193,4 +332,141 @@ function (Builder $query) use ($airfield) {
+
+        // Find the stand to which the aircraft is closest
+        $selectedStand = null;
+        $selectedStandDistance = PHP_INT_MAX;
+
+        foreach ($this->getAllStands() as $stand) {
+            $distanceFromStand = $stand->coordinate->getDistance($aircraft->latLong, new Haversine());
+
+            if (
+                $this->standOccupied($aircraft, $stand) &&
+                $distanceFromStand < $selectedStandDistance
+            ) {
+                $selectedStand = $stand;
+                $selectedStandDistance = $distanceFromStand;
+            }
+        }
+
+        // If there's a stand that's viable, usurp any assignments and occupy it.
+        if ($selectedStand) {
+            $this->usurpStand($aircraft, $selectedStand);
+            $aircraft->occupiedStand()->sync([$selectedStand->id]);
+        }
+
+        return $selectedStand;
+    }
+
+    /*
+     * Delete any stand assignment that isn't for the given aircraft
+     */
+    private function usurpStand(NetworkAircraft $aircraft, Stand $stand)
+    {
+        $conflictingAssignment = StandAssignment::where('callsign', '<>', $aircraft->callsign)
+            ->where('stand_id', $stand->id)
+            ->first();
+
+        if ($conflictingAssignment) {
+            $this->deleteStandAssignment($conflictingAssignment);
+        }
+    }
+
+    private function standOccupied(NetworkAircraft $aircraft, Stand $stand): bool
+    {
+        $distanceFromStand = $stand->coordinate->getDistance($aircraft->latLong, new Haversine());
+        return $distanceFromStand < self::MAX_OCCUPANCY_DISTANCE_METERS;
+    }
+
+    private function aircraftCanOccupyStand(NetworkAircraft $aircraft): bool
+    {
+        return $aircraft->altitude < self::MAX_OCCUPANCY_ALTITUDE
+            && $aircraft->groundspeed < self::MAX_OCCUPANCY_SPEED;
+    }
+
+    /**
+     * Use the stand assignment rules to allocate a stand for a given aircraft
+     */
+    public function allocateStandForAircraft(NetworkAircraft $aircraft): ?StandAssignment
+    {
+        if (!$this->shouldAllocateStand($aircraft)) {
+            return null;
+        }
+
+        foreach ($this->allocators as $allocator) {
+            if ($allocation = $allocator->allocate($aircraft)) {
+                event(new StandAssignedEvent($allocation));
+                return $allocation;
+            }
+        }
+
+        return null;
+    }
+
+    private function shouldAllocateStand(NetworkAircraft $aircraft): bool
+    {
+        return !StandAssignment::where('callsign', $aircraft->callsign)->exists() &&
+            ($arrivalAirfield = Airfield::where('code', $aircraft->planned_destairport)->first()) !== null &&
+            $aircraft->groundspeed &&
+            ($aircraftType = Aircraft::where('code', $aircraft->aircraftType)->first()) &&
+            $aircraftType->allocate_stands &&
+            $this->getTimeFromAirfieldInMinutes($aircraft, $arrivalAirfield) < self::ASSIGN_STAND_MINUTES_BEFORE;
+    }
```

Comment:
This is really hard to decifer. I understand it is deciding whether we should allocate a stand but its difficult to tell what the criteria is.

Reply:
Added a comment to clarify